Project-Management/CoreTeam/workspace-notes.md

```diff
@@ -0,0 +1,43 @@
+# Workspace
+
+## Issues with the current workspace
+
+Focussing mainly on `MatrixWorkspace`.
+
+1. Constructed iteratively, i.e., workspaces start their life in an incomplete and dysfunctional state, vital information is added only later (size, instrument, spectrum to detector mapping).
+1. No encapsulation for data it contains, thus violating invariants:
+   - Access to X, Y, and E makes breaking or abusing internal data easy.
+   - Access to spectrum numbers on `ISpectrum` prevents validation (e.g., duplicates).
+   - Detector IDs can be set arbitrarily on `ISpectrum`, no check if this is in the instrument.
+1. `MatrixWorkspace` is meant to store histograms, but is used for many other purposes:
+   - Single values.
+   - Single data points in each histogram. Leads to performance problems due to massive overhead from using histograms for this purpose, e.g., at ILL.
+   - Masking.
+   - Grouping.
+   - HKL values.
+1. Not enough workspaces types (thus the aforementioned misappropriation of `MatrixWorkspace`).
+   - In the current way workspaces and algorithms are built, adding new workspaces types is not very useful since existing algorithms will not work for them.
+1. The base class of `MatrixWorkspace`, `ExperimentInfo` is a complicated and holds a lot of complexity related to setting the instrument and loading parameter maps.
+   - Provides a detector grouping mechanism based on externally set grouping (i.e., *not* the grouping used by `MatrixWorkspace`), used by `MDWorkspace`.
+1. Very complicated to copy / create output workspaces.
```

Comment:
No way to pickle/serialize a workspace in python. This is increasingly required for ECP/Data Reduction interop.

Reply:
True. How does this relate to passing workspaces between MPI ranks too?

Reply:
This is still unclear but the way I see things currently we will usually *not* pass complete workspaces between MPI ranks but just, e.g., a vector of histograms. All other parts of the workspace are available on all MPI ranks.